internal/ppa/ppa.go

```diff
@@ -0,0 +1,377 @@
+}
+
+func (p *BasePPA) createTmpGPGDir(basePath string) (string, error) {
+	tmpGPGDir := filepath.Join(basePath, "tmp", "ubuntu-image-gpg")
+	err := osMkdirAll(tmpGPGDir, 0755)
+	if err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("Error creating temp dir for gpg imports: %s", err.Error())
+	}
+	return tmpGPGDir, nil
+}
+
+type PPA struct {
+	PPAPrivateInterface
+}
+
+// Add adds the PPA to the sources.list.d directory and import the signing key.
+func (p *PPA) Add(basePath string, debug bool) error {
+	sourcesListD := filepath.Join(basePath, sourcesListDPath)
+	err := osMkdirAll(sourcesListD, 0755)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Failed to create apt sources.list.d: %s", err.Error())
+	}
+
+	err = p.ImportKey(basePath, debug)
+	if err != nil {
+		return fmt.Errorf("Error retrieving signing key for ppa \"%s\": %s",
+			p.FullName(), err.Error())
+	}
+
+	var ppaIO *os.File
+	ppaFile := filepath.Join(sourcesListD, p.FileName())
+	ppaIO, err = osOpenFile(ppaFile, os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("Error creating %s: %s", ppaFile, err.Error())
+	}
+	defer ppaIO.Close()
+
+	content, err := p.FileContent()
+	if err != nil {
+		return err
+	}
+
+	_, err = ppaIO.Write([]byte(content))
+	if err != nil {
+		return fmt.Errorf("unable to write ppa file %s: %w", ppaFile, err)
+	}
+
+	return nil
+}
+
+// LegacyPPA implements behaviors to manage PPA the legacy way, specifically:
+// - write in a sources.list file
+// - manage signing key with gpg in /etc/apt/trusted.gpg.d
+type LegacyPPA struct {
+	BasePPA
+}
+
+func (p *LegacyPPA) FileName() string {
+	return fmt.Sprintf("%s-ubuntu-%s-%s.list", p.user(), p.name(), p.Series)
+}
+
+func (p *LegacyPPA) FileContent() (string, error) {
+	return fmt.Sprintf("deb %s %s main", p.url(), p.Series), nil
+}
+
+func (p *LegacyPPA) ImportKey(basePath string, debug bool) error {
+	return p.BasePPA.importKey(basePath, p.FileName(), false, debug)
+}
+
+func (p *LegacyPPA) Remove(basePath string) error {
+	err := p.removePPAFile(basePath, p.FileName())
+	if err != nil {
+		return err
+	}
+
+	trustedGPGD := filepath.Join(basePath, trustedGPGDPath)
+	keyFileName := strings.Replace(p.FileName(), ".list", ".gpg", 1)
+	keyFilePath := filepath.Join(trustedGPGD, keyFileName)
+
+	err = osRemove(keyFilePath)
+	if err != nil {
+		return fmt.Errorf("Error removing %s: %s", keyFilePath, err.Error())
+	}
+
+	return nil
+}
+
+// Deb822PPA implements behaviors to manage PPA in the deb822 format, specifically:
+// - write in a <ppa>.sources file, in the deb822 format
+// - embed the signing key in the file itself
+type Deb822PPA struct {
+	BasePPA
+}
+
+func (p *Deb822PPA) FileName() string {
+	return fmt.Sprintf("%s-ubuntu-%s-%s.sources", p.user(), p.name(), p.Series)
+}
+
+func (p *Deb822PPA) FileContent() (string, error) {
+	key, err := p.formatKey(p.BasePPA.signingKey)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("X-Repolib-Name: %s\nTypes: deb\n"+
```

Comment:
FWIW I don't know what that `X-Repolib-Name` is but I'd prefer it just before `Signed-By` otherwise this looks a bit weird compared to other files. Note that software-properties doesn't add one.

Reply:
After reading the doc again I spotted I was reading the doc of a tool working with deb822 and that this field is specific to a tool. I will remove it.